source/1_engine/rex_engine/include/rex_engine/state_controller.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "rex_engine/types.h"
+
+#include <rex_std/type_traits.h>
+
+namespace rex
+{
+  template <typename TState, rsl::enable_if_t<rsl::is_enum_v<TState>, bool> = true>
+  class StateController
+  {
+  public:
+    StateController()
+        : m_state(0)
+    {
+    }
+    explicit StateController(TState state)
+        : m_state(static_cast<u32>(state))
+    {
+    }
+
+    void add_state(TState state)
+    {
+      m_state = m_state | static_cast<u32>(state);
+    }
+    void remove_state(TState state)
+    {
+      m_state = m_state & ~static_cast<u32>(state);
+    }
+
+    bool has_state(TState state) const
+    {
+      return (m_state & static_cast<u32>(state)) != 0u;
+    }
+
+  private:
+    u32 m_state;
```

Comment:
You told me I should move it to something that can be used in a more generic way, now I did and now you tell me I over engineer. Make up your mind

Reply:
can you remind me about my comment about making it more generic?
yes, something like this should be generic, but it doesn't have to be in a class imo.
like I said, it can only be in 1 state, so you just update the enum

Reply:
Previously this was defined within the window and you told me that this "state definition" will be useful in some library manner as we'll probably have more states ( eg: application state ). So I moved it into a class that holds a state of something. Both the window and the application can have multiple flags enabled not a single one.  For example the application can be running but marked for destroy the window can be maximized and paused.

Reply:
if an application is marked for destroy, the running state is implied, anyway, the multiple states is less of an issue here imo.
I do remember saying that we'll have more states and when I said it should be moved in some library I was talking more about just functions, not really into a class.
You can keep it in a class if you want to

Reply:
The functions will also store some global state probably this is just encapsulated into a class, I think that's just a way of handling it I prefer classes over functions

Reply:
the logic was in Window, not separate into its own system.
it's fine if you want to keep the class

Reply:
The same logic was also in Application which is why I changed it as well but oke I'll keep it.

Reply:
you're still using an int as a type to store the state instead of the enum template type that's passed in.